// angular-circle-counter/esm5/src/app/counter/counter.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { NgModule, CUSTOM_ELEMENTS_SCHEMA } from '@angular/core';
import { CounterComponent } from './counter.component';
import { CommonModule } from '@angular/common';
var CounterModule = /** @class */ (function () {
    function CounterModule() {
    }
    CounterModule.decorators = [
        { type: NgModule, args: [{
                    declarations: [
                        CounterComponent
                    ],
                    imports: [
                        CommonModule
                    ],
                    schemas: [CUSTOM_ELEMENTS_SCHEMA],
                    exports: [
                        CounterComponent
                    ]
                },] },
    ];
    return CounterModule;
}());
export { CounterModule };
function CounterModule_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    CounterModule.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    CounterModule.ctorParameters;
}
